### PR TITLE
[MM-45761] - Telemetry: Track how many users end up in purchase/pricing modal/notify admin

### PR DESCRIPTION
**Pull request by AGMETEOR**
#### Summary
This PR adds the ability to know in which component the purchase modal opening was initiated from

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-45761
https://mattermost.atlassian.net/browse/MM-45014

#### Related PRs
https://github.com/mattermost/focalboard/pull/3668
https://github.com/mattermost/mattermost-plugin-calls/pull/162
https://github.com/mattermost/mattermost-plugin-playbooks/pull/1404

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Review by marianunez — COMMENTED** [review]
Changes LGTM. - 

Based on this comment https://mattermost.atlassian.net/browse/MM-45761?focusedCommentId=132079, 

Is there a way we can track when the purchase modal was opened directly from the link we provide in mm.com/pricing?

File: components/admin_console/billing/billing_subscriptions/index.tsx [hunk 1]
```diff
@@ -76,11 +77,11 @@ const BillingSubscriptions = () => {
     const product = useSelector(getSubscriptionProduct);
 
     const openPricingModal = useOpenPricingModal();
+    const openCloudPurchaseModal = useOpenCloudPurchaseModal({});
 
     // show the upgrade section when is a free tier customer
-    const onUpgradeMattermostCloud = () => {
-        trackEvent('cloud_admin', 'click_upgrade_mattermost_cloud');
-        openPricingModal();
```
Comment (AGMETEOR) [inline components/admin_console/billing/billing_subscriptions/index.tsx:83]: @marianunez I had missed it. Added telemetry for it now

**Comment by johndavidlugtu** [comment]
@AGMETEOR please add @sri-byte as a reviewer -- he doesn't need to approve, but good to have visibility

**Review by neallred — APPROVED** [review]
LGTM, thanks @AGMETEOR !

File: components/admin_console/billing/billing_subscriptions/cloud_trial_banner.tsx [hunk 2]
```diff
@@ -92,6 +92,7 @@ const CloudTrialBanner = ({trialEndDate}: Props): JSX.Element | null => {
                 <UpgradeLink
                     buttonText={formatMessage({id: 'admin.subscription.cloudTrialCard.upgrade', defaultMessage: 'Upgrade'})}
                     styleButton={true}
+                    telemetryInfo='admin_console_billing_subscriptions_cloud_trial_banner'
```
Comment (marianunez) [inline components/admin_console/billing/billing_subscriptions/cloud_trial_banner.tsx:95]:
```suggestion
                    telemetryInfo='billing_subscriptions_cloud_trial_banner'
```

File: components/admin_console/billing/billing_subscriptions/contact_sales_card.tsx [hunk 3]
```diff
@@ -159,7 +159,7 @@ const ContactSalesCard = (props: Props) => {
                 {(!isFreeTrial && subscriptionPlan !== CloudProducts.ENTERPRISE && subscriptionPlan !== CloudProducts.LEGACY) &&
                     <button
                         type='button'
-                        onClick={onUpgradeMattermostCloud}
+                        onClick={() => onUpgradeMattermostCloud('admin_console_billing_contact_sales_card_upgrade_now_button')}
```
Comment (marianunez) [inline components/admin_console/billing/billing_subscriptions/contact_sales_card.tsx:162]: This Contact Sales card I believe it changes to Upgrade (with no reference to sales) when not in Enterprise. Removed the reference to Contact Sales as is misleading

```suggestion
                        onClick={() => onUpgradeMattermostCloud('admin_console_subscription_card_upgrade_now_button')}
```

File: components/admin_console/billing/billing_subscriptions/index.tsx [hunk 4]
```diff
@@ -108,11 +109,11 @@ const BillingSubscriptions = () => {
         pageVisited('cloud_admin', 'pageview_billing_subscription');
 
         if (actionQueryParam === 'show_purchase_modal') {
-            onUpgradeMattermostCloud();
+            onUpgradeMattermostCloud('admin_console_billing_billing_subscriptions_action_query_param');
```
Comment (marianunez) [inline components/admin_console/billing/billing_subscriptions/index.tsx:112]: Is this when it's called from the link? Lets update `action_query_param` to something more understandable for analytics. Added a suggestion but open to other options
```suggestion
            onUpgradeMattermostCloud('billing_subscriptions_external_direct_link');
```

File: components/admin_console/billing/billing_subscriptions/index.tsx [hunk 5]
```diff
@@ -108,11 +109,11 @@ const BillingSubscriptions = () => {
...
         }
 
         if (actionQueryParam === 'show_pricing_modal') {
-            openPricingModal();
+            openPricingModal({trackingLocation: 'admin_console_billing_billing_subscriptions_action_query_param'});
```
Comment (marianunez) [inline components/admin_console/billing/billing_subscriptions/index.tsx:116]: Same as above

File: components/admin_console/billing/billing_subscriptions/limits.tsx [hunk 6]
```diff
@@ -243,7 +243,7 @@ const Limits = (props: Props): JSX.Element | null => {
                 {subscriptionProduct.sku === CloudProducts.STARTER && (
                     <>
                         <button
-                            onClick={openPricingModal}
+                            onClick={() => openPricingModal({trackingLocation: 'admin_console_billing_billing_subscriptions_limits'})}
```
Comment (marianunez) [inline components/admin_console/billing/billing_subscriptions/limits.tsx:246]:
```suggestion
                            onClick={() => openPricingModal({trackingLocation: 'billing_subscriptions_limits_dashboard'})}
```

File: components/admin_console/billing/billing_summary/billing_summary.tsx [hunk 7]
```diff
@@ -72,7 +72,7 @@ export const upgradeFreeTierMattermostCloud = (onUpgradeMattermostCloud: () => v
         </div>
         <button
             type='button'
-            onClick={onUpgradeMattermostCloud}
+            onClick={() => onUpgradeMattermostCloud('admin_console_billing_summary_upgrade_legacy_free_tier_upgrade_now_button')}
```
Comment (marianunez) [inline components/admin_console/billing/billing_summary/billing_summary.tsx:75]:
```suggestion
            onClick={() => onUpgradeMattermostCloud('billing_summary_legacy_free_tier_upgrade_button')}
```

File: components/admin_console/billing/billing_summary/billing_summary.tsx [hunk 8]
```diff
@@ -129,7 +129,7 @@ export const freeTrial = (onUpgradeMattermostCloud: () => void, daysLeftOnTrial:
         </div>
         <button
             type='button'
-            onClick={onUpgradeMattermostCloud}
+            onClick={() => onUpgradeMattermostCloud('admin_console_billing_summary_free_trial_upgrade_now_button')}
```
Comment (marianunez) [inline components/admin_console/billing/billing_summary/billing_summary.tsx:132]:
```suggestion
            onClick={() => onUpgradeMattermostCloud('billing_summary_free_trial_upgrade_button')}
```

File: components/admin_console/billing/billing_summary/upsell_card.tsx [hunk 9]
```diff
@@ -163,7 +163,7 @@ export const UpgradeToProfessionalCard = () => {
                 id: t('admin.billing.subscriptions.billing_summary.upgrade_professional.cta'),
                 defaultMessage: 'Upgrade',
             }}
-            ctaAction={openPurchaseModal}
+            ctaAction={() => openPurchaseModal({trackingLocation: 'admin_console_billing_summary_upsell_card_upgrade_professional_card'})}
```
Comment (marianunez) [inline components/admin_console/billing/billing_summary/upsell_card.tsx:166]:
```suggestion
            ctaAction={() => openPurchaseModal({trackingLocation: 'billing_summary_upsell_professional_card'})}
```

File: components/admin_console/billing/plan_details/plan_details.tsx [hunk 10]
```diff
@@ -100,7 +100,7 @@ export const PlanDetailsTopElements = ({
 
     const viewPlansButton = (
         <button
-            onClick={openPricingModal}
+            onClick={() => openPricingModal({trackingLocation: 'admin_console_billing_plan_details_view_plans'})}
```
Comment (marianunez) [inline components/admin_console/billing/plan_details/plan_details.tsx:103]:
```suggestion
            onClick={() => openPricingModal({trackingLocation: 'billing_plan_details_view_plans'})}
```

File: components/admin_console/feature_discovery/feature_discovery.tsx [hunk 11]
```diff
@@ -87,6 +87,9 @@ export default class FeatureDiscovery extends React.PureComponent<Props, State>
         this.props.actions.openModal({
             modalId: ModalIdentifiers.CLOUD_PURCHASE,
             dialogType: PurchaseModal,
+            dialogProps: {
+                callerCTA: 'click_subscribe_from_feature_discovery',
```
Comment (marianunez) [inline components/admin_console/feature_discovery/feature_discovery.tsx:91]:
```suggestion
                callerCTA: 'feature_discovery_subscribe_button',
```

File: components/center_message_lock/index.tsx [hunk 12]
```diff
@@ -93,7 +93,7 @@ export default function CenterMessageLock(props: Props) {
     let cta = (
         <button
             className='btn btn-primary'
-            onClick={notifyAdmin}
+            onClick={(e) => notifyAdmin(e, 'center_message_lock')}
```
Comment (marianunez) [inline components/center_message_lock/index.tsx:96]: What does CenterMessageLock refer to? Is this is they are over the limits in posts?

File: components/cloud_usage_modal/lhs_nearing_limit_modal.tsx [hunk 13]
```diff
@@ -36,7 +36,7 @@ export default function LHSNearingLimitsModal() {
             id: t('workspace_limits.modals.view_plans'),
             defaultMessage: 'View plans',
         },
-        onClick: openPricingModal,
+        onClick: () => openPricingModal({trackingLocation: 'cloud_usage_modal_lhs_nearing_limit_modal'}),
```
Comment (marianunez) [inline components/cloud_usage_modal/lhs_nearing_limit_modal.tsx:39]:
```suggestion
        onClick: () => openPricingModal({trackingLocation: 'cloud_usage_lhs_nearing_limit_modal'}),
```

**Review by marianunez — COMMENTED** [review]
Great stuff @AGMETEOR! That's a lot of places that it is called to cover. Added some suggestions to reduce and/or clarify the caller info names.  Also open to additional input on the names from @sri-byte @johndavidlugtu 

What do you recommend to test this? Could we add some instructions for @yasserfaraazkhan to validate here?

File: components/center_message_lock/index.tsx (hunk 12 above)
Comment (AGMETEOR) [reply components/center_message_lock/index.tsx:96]: @marianunez yes that's correct

Comment (marianunez) [reply components/center_message_lock/index.tsx:96]: Ok, so maybe we can do something like:
```suggestion
            onClick={(e) => notifyAdmin(e, 'center_channel_posts_over_limit_banner')}
```
?

**Comment by AGMETEOR** [comment]
> Great stuff @AGMETEOR! That's a lot of places that it is called to cover. Added some suggestions to reduce and/or clarify the caller info names. Also open to additional input on the names from @sri-byte @johndavidlugtu
> 
> What do you recommend to test this? Could we add some instructions for @yasserfaraazkhan to validate here?

@yasserfaraazkhan I tested with a console.log statement down in the trackEvent method

File: components/center_message_lock/index.tsx [hunk 14]
```diff
@@ -116,7 +116,7 @@ export default function CenterMessageLock(props: Props) {
                         href='#'
                         onClick={(e: React.MouseEvent) => {
                             e.preventDefault();
-                            openPricingModal();
+                            openPricingModal({trackingLocation: 'center_message_lock'});
```
Comment (marianunez) [inline components/center_message_lock/index.tsx:119]:
```suggestion
                            openPricingModal({trackingLocation: 'center_channel_posts_over_limit_banner'});
```

File: components/center_message_lock/index.tsx [hunk 15]
```diff
@@ -128,7 +128,7 @@ export default function CenterMessageLock(props: Props) {
         cta = (
             <button
                 className='btn is-admin'
-                onClick={openPricingModal}
+                onClick={() => openPricingModal({trackingLocation: 'center_message_lock'})}
```
Comment (marianunez) [inline components/center_message_lock/index.tsx:131]:
```suggestion
                onClick={() => openPricingModal({trackingLocation: 'center_channel_posts_over_limit_banner''})}
```

File: components/custom_open_pricing_modal_post_renderer/index.tsx [hunk 16]
```diff
@@ -44,7 +44,7 @@ export default function OpenPricingModalPost(props: {post: Post}) {
                     style={style}
                 >
                     <button
-                        onClick={openPricingModal}
+                        onClick={() => openPricingModal({trackingLocation: 'system_bot_notify_admin_message_view_upgrade_options'})}
```
Comment (marianunez) [inline components/custom_open_pricing_modal_post_renderer/index.tsx:47]:
```suggestion
                        onClick={() => openPricingModal({trackingLocation: 'notify_admin_message_view_upgrade_options'})}
```

**Review by marianunez — APPROVED** [review]

**Review by yasserfaraazkhan — APPROVED** [review]
Thank you Allan.

**Review by sri-byte — APPROVED** [review]

**Comment by cwarnermm** [comment]
@AGMETEOR - Can you summarize the changes needed for the [telemetry](https://docs.mattermost.com/manage/telemetry.html) documentation page, please?
